### PR TITLE
Minor fixes for hashed dictionary

### PR DESCRIPTION
**Pull request by vdimir**
### Changelog category (leave one):

- Not for changelog (changelog entry is not required)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
...

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by robot-ch-test-poll2** [comment]
*This is an automated comment for commit 4a7ad15a9a7cf9094ff44a010ae08386036ada9a with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/60310/4a7ad15a9a7cf9094ff44a010ae08386036ada9a/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/60310/4a7ad15a9a7cf9094ff44a010ae08386036ada9a/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

File: src/Dictionaries/HashedArrayDictionary.cpp [hunk 1]
```diff
@@ -822,10 +822,12 @@ void HashedArrayDictionary<dictionary_key_type, sharded>::loadData()
         if (parallel_loader)
             parallel_loader->finish();
 
-        LOG_DEBUG(getLogger("HashedArrayDictionary"),
-            "Finished {}reading {} blocks with {} rows from pipeline in {:.2f} sec and inserted into hashtable in {:.2f} sec",
+        LOG_DEBUG(log,
+            "Finished {}reading {} blocks with {} rows to dictionary {} from pipeline in {:.2f} sec and inserted into hashtable in {:.2f} sec",
```
Comment (novikd) [inline src/Dictionaries/HashedArrayDictionary.cpp:826]:
```suggestion
            "Finished {}reading {} blocks with {} rows to dictionary `{}` from pipeline in {:.2f} sec and inserted into hashtable in {:.2f} sec",
```

File: src/Dictionaries/HashedDictionaryParallelLoader.h [hunk 2]
```diff
@@ -46,12 +46,13 @@ class HashedDictionaryParallelLoader : public boost::noncopyable
...
     explicit HashedDictionaryParallelLoader(DictionaryType & dictionary_)
         : dictionary(dictionary_)
+        , dictionary_name(dictionary.getFullName())
         , shards(dictionary.configuration.shards)
         , pool(CurrentMetrics::HashedDictionaryThreads, CurrentMetrics::HashedDictionaryThreadsActive, CurrentMetrics::HashedDictionaryThreadsScheduled, shards)
         , shards_queues(shards)
     {
         UInt64 backlog = dictionary.configuration.shard_load_queue_backlog;
-        LOG_TRACE(dictionary.log, "Will load the dictionary using {} threads (with {} backlog)", shards, backlog);
+        LOG_TRACE(dictionary.log, "Will load the {} dictionary using {} threads (with {} backlog)", dictionary_name, shards, backlog);
```
Comment (novikd) [inline src/Dictionaries/HashedDictionaryParallelLoader.h:55]:
```suggestion
        LOG_TRACE(dictionary.log, "Will load the dictionary `{}` using {} threads (with {} backlog)", dictionary_name, shards, backlog);
```

File: src/Dictionaries/HashedDictionaryParallelLoader.h [hunk 3]
```diff
@@ -98,7 +100,7 @@ class HashedDictionaryParallelLoader : public boost::noncopyable
         Stopwatch watch;
         pool.wait();
         UInt64 elapsed_ms = watch.elapsedMilliseconds();
-        LOG_TRACE(dictionary.log, "Processing the tail took {}ms", elapsed_ms);
+        LOG_TRACE(dictionary.log, "Processing the tail of dictionary {} took {}ms", dictionary_name, elapsed_ms);
```
Comment (novikd) [inline src/Dictionaries/HashedDictionaryParallelLoader.h:103]:
```suggestion
        LOG_TRACE(dictionary.log, "Processing the tail of dictionary `{}` took {}ms", dictionary_name, elapsed_ms);
```

**Review by novikd — APPROVED** [review]

File: src/Dictionaries/HashedArrayDictionary.cpp (hunk 1 above)
Comment (vdimir) [reply src/Dictionaries/HashedArrayDictionary.cpp:826]: It returns name for logs from https://github.com/ClickHouse/ClickHouse/blob/4a7ad15a9a7cf9094ff44a010ae08386036ada9a/src/Interpreters/StorageID.cpp#L60-L66, so ticks are excessive here

File: src/Dictionaries/HashedArrayDictionary.cpp [hunk 4]
```diff
@@ -969,6 +971,9 @@ void registerDictionaryArrayHashed(DictionaryFactory & factory)
 
         HashedArrayDictionaryStorageConfiguration configuration{require_nonempty, dict_lifetime, static_cast<size_t>(shards)};
 
+        if (source_ptr->hasUpdateField() && shards > 1)
```
Comment (vdimir) [inline src/Dictionaries/HashedArrayDictionary.cpp:974]: We have the same for `hashed` dictionary, but it was missed for `hashed_array`. If we specify both `update_field` and `SHARDS` dictionary will not work properly and will return wrong results.

**Comment by vdimir** [comment]
> Integration tests (asan, analyzer) [2/6] — fail: 1, passed: 375 [Details](https://s3.amazonaws.com/clickhouse-test-reports/60310/d62b8a5d4a08930e932cc64fcc14528545622a13/integration_tests__asan__analyzer__[2_6].html)

? `test_failed_mutations/test.py::test_exponential_backoff_with_merge_tree`

> Integration tests (tsan) [3/6] — fail: 1, passed: 335 [Details](https://s3.amazonaws.com/clickhouse-test-reports/60310/d62b8a5d4a08930e932cc64fcc14528545622a13/integration_tests__tsan__[3_6].html)

https://github.com/ClickHouse/ClickHouse/issues/60347

> Stateless tests (aarch64) — fail: 1, passed: 6099, skipped: 33 [Details](https://s3.amazonaws.com/clickhouse-test-reports/60310/d62b8a5d4a08930e932cc64fcc14528545622a13/stateless_tests__aarch64_.html)

? `02125_many_mutations`

> Stateless tests (debug) [1/5] — fail: 1, passed: 1236, skipped: 15 [Details](https://s3.amazonaws.com/clickhouse-test-reports/60310/d62b8a5d4a08930e932cc64fcc14528545622a13/stateless_tests__debug__[1_5].html)

? `01459_manual_write_to_replicas_quorum` timeout

**Comment by clickgapai** [comment]
Hi @vdimir @novikd — while reviewing this PR I found the following:

- **Coverage gap:** New `HASHED_ARRAY` rejection of `SHARDS > 1` + `UPDATE_FIELD` is untested — https://github.com/ClickHouse/ClickHouse/pull/103884

Could you review the test PR and add the `can be tested` label if it looks good?
Happy to discuss — close anything that's wrong or already addressed.

**Pull request merged**